programs/drift/src/controller/orders.rs

```diff
@@ -2820,6 +2865,118 @@ pub fn can_reward_user_with_perp_pnl(user: &mut Option<&mut User>, market_index:
     }
 }
 
+pub fn attempt_burn_user_lp_shares_for_risk_reduction(
+    state: &State,
+    user: &mut User,
+    margin_calc: MarginCalculation,
+    user_key: &Pubkey,
+    perp_market_map: &PerpMarketMap,
+    spot_market_map: &SpotMarketMap,
+    oracle_map: &mut OracleMap,
+    clock: &Clock,
+) -> DriftResult {
+    let now = clock.unix_timestamp;
+    // attempt to burn lp shares if user has a custom margin ratio set and its breached with orders
+    if !margin_calc.positions_meets_margin_requirement()? {
+        let time_since_last_liquidity_change: i64 =
+            now.safe_sub(user.last_add_perp_lp_shares_ts)?;
+        // avoid spamming update if orders have already been set
+        if time_since_last_liquidity_change >= state.lp_cooldown_time.cast()? {
+            let set_reduce_only_orders = true; // todo
+            for position_index in 0..user.perp_positions.len() {
+                let market_index = user.perp_positions[position_index].market_index;
+                _burn_user_lp_shares_for_risk_reduction(
+                    state,
+                    user,
+                    user_key,
+                    market_index,
+                    perp_market_map,
+                    spot_market_map,
+                    oracle_map,
+                    clock,
+                    set_reduce_only_orders,
+                )?;
+            }
+            user.last_add_perp_lp_shares_ts = now;
+        }
+    }
+
+    Ok(())
+}
+
+pub fn _burn_user_lp_shares_for_risk_reduction(
+    state: &State,
+    user: &mut User,
+    user_key: &Pubkey,
+    market_index: u16,
+    perp_market_map: &PerpMarketMap,
+    spot_market_map: &SpotMarketMap,
+    oracle_map: &mut OracleMap,
+    clock: &Clock,
+    set_reduce_only_orders: bool,
+) -> DriftResult {
+    let perp_position = user.get_perp_position_mut(market_index)?;
+    if perp_position.is_lp() {
+        let lp_shares = perp_position.lp_shares;
+        let market_index = perp_position.market_index;
+        {
+            let market = perp_market_map.get_ref(&market_index)?;
+            let oracle_price_data = oracle_map.get_price_data(&market.amm.oracle)?;
+
+            let oracle_price = if market.status == MarketStatus::Settlement {
+                market.expiry_price
+            } else {
+                oracle_price_data.price
+            };
+
+            let (position_delta, pnl) = burn_lp_shares(
+                perp_position,
+                perp_market_map.get_ref_mut(&market_index)?.deref_mut(),
```

Comment:
you cant do this after get perp market above, will panic

it should just get it mutable above and pass it ehre